### PR TITLE
[Issue 864] save module weights, load (all / part) module weights

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

load and save weights.

This is useful for finetuning, If you need to load weights into a different architecture (with some layers in common), for instance for fine-tuning or transfer-learning, you can load weights by layer name.

## How was this patch tested?

unit test

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/864

### CODE REVIEW

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by qiuxin2012** [comment]
LGTM

**Comment by SeaOfOcean** [comment]
that's a problem, I think this is the same problem for model.save

**Comment by qiuxin2012** [comment]
Could you try save/loadTorch?

**Comment by SeaOfOcean** [comment]
There is SerialVersionUID in Tensor, so adding fields will not affect loading old files. Having tried and it works

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 1]
```diff
@@ -432,6 +432,75 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+  }
+
+  /**
+   * load pretrained weights and bias to current module
+   * @param weightPath file to store weights and bias
+   * @param matchAll whether to match all layers' weights and bias,
+   *                 if not, only load existing pretrained weights and bias
+   * @return current module
+   */
+  def loadWeights(weightPath: String, matchAll: Boolean = true): this.type = {
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: where do you use matchAll

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 2]
```diff
@@ -432,6 +432,75 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+    val srcParameter = File.load[Table](weightPath)
+    val targetParameter = getParametersTable()
+    copyWeights(targetParameter, srcParameter)
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: perhaps here miss param "matchAll"

Comment (SeaOfOcean) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: yes

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala [hunk 3]
```diff
@@ -432,6 +432,75 @@ abstract class AbstractModule[A <: Activity: ClassTag, B <: Activity: ClassTag,
...
+
+  /**
+   * copy weights from another model, mapping by layer name
+   * @param srcModel model to copy from
+   * @param matchAll whether to match all layers' weights and bias,
+   * @return current module
+   */
+  def loadModelWeights(srcModel: Module[Float], matchAll: Boolean = true): this.type = {
+    val srcParameters = srcModel.getParametersTable()
+    val parameterTable = getParametersTable()
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/AbstractModule.scala]: similar to "loadweights" function, name "targetParameter" is better

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by zhangxiaoli73** [comment]
LGTM

**Pull request merged**